python_modules/libraries/dagster-components/dagster_components/lib/custom_component.py

```diff
@@ -0,0 +1,89 @@
+import inspect
+import os
+import textwrap
+from abc import abstractmethod
+from pathlib import Path
+from typing import TYPE_CHECKING, Any, Mapping, Optional
+
+import click
+from dagster._utils import snakecase
+from pydantic import BaseModel
+
+from dagster_components.core.component import (
+    Component,
+    ComponentGenerateRequest,
+    ComponentLoadContext,
+    component,
+)
+from dagster_components.generate import generate_custom_component_yaml
+
+if TYPE_CHECKING:
+    from dagster._core.definitions.definitions_class import Definitions
+
+
+def custom_component_template():
+    from dagster import Definitions
+    from pydantic import BaseModel
+
+    from dagster_components import ComponentLoadContext, component
+    from dagster_components.lib.custom_component import CustomComponent
+
+    class ClassNamePlaceholderComponentSchema(BaseModel): ...
+
+    @component(name="component_type_name_placeholder")
+    class ClassNamePlaceholderComponent(CustomComponent):
+        params_schema = ClassNamePlaceholderComponentSchema
+
+        def build_defs(self, context: ComponentLoadContext) -> Definitions:
+            return Definitions()
+
+        @classmethod
+        def load(cls, context: ComponentLoadContext) -> "ClassNamePlaceholderComponent":
+            loaded_params = context.load_params(cls.params_schema)
+            assert loaded_params  # silence linter complaints
+            return cls()
+
+
+CLASSNAME_PLACEHODLER = "ClassNamePlaceholder"
+COMPONENT_TYPE_NAME_PLACEHOLDER = "component_type_name_placeholder"
+
+
+class GenerateCustomComponentParams(BaseModel):
+    class_name: str
+
+    @staticmethod
+    @click.command
+    @click.option("--class-name", "- ", type=click.STRING)
+    def cli(class_name: str) -> "GenerateCustomComponentParams":
+        return GenerateCustomComponentParams(class_name=class_name)
```

Comment:
`cli` here can be removed with the new CLI option exposure system

Reply:
right. that is not merged yet, correct?